pkg/virt-controller/watch/vm.go

```diff
@@ -428,41 +429,88 @@ func (c *VMController) handleDataVolumes(vm *virtv1.VirtualMachine, dataVolumes
 	return ready, nil
 }
 
-func (c *VMController) startStop(vm *virtv1.VirtualMachine, vmi *virtv1.VirtualMachineInstance) error {
-	log.Log.Object(vm).V(4).Infof("Start the VirtualMachineInstance: %t", vm.Spec.Running)
+func (c *VMController) startStop(vm *virtv1.VirtualMachine, vmi *virtv1.VirtualMachineInstance) (bool, error) {
+	RunStrategy := getVirtualMachineRunStrategy(vm)
+	log.Log.Object(vm).V(4).Infof("VirtualMachine RunStrategy: %s", RunStrategy)
 
-	if vm.Spec.Running == true {
+	if vm.Status.RestartRequest != nil {
+		if vm.Status.RestartRequest.UUID != nil {
+			log.Log.Object(vm).V(4).Infof("VirtualMachine RestartRequest UUID: %s", *vm.Status.RestartRequest.UUID)
+		} else {
+			log.Log.Object(vm).V(4).Infof("VirtualMachine RestartRequest with no UUID")
+		}
+
+		shouldDelete := false
+		shouldStart := false
 		if vmi != nil {
-			if vmi.IsFinal() {
-				// The VirtualMachineInstance can fail od be finished. The job of this controller
-				// is keep the VirtualMachineInstance running, therefore it restarts it.
-				// restarting VirtualMachineInstance by stopping it and letting it start in next step
+			if vm.Status.RestartRequest.UUID != nil && vmi.UID == *vm.Status.RestartRequest.UUID {
+				shouldDelete = true
+			}
+		}
+		if shouldDelete || vmi == nil {
+			shouldStart = true
+		}
+
+		if RunStrategy != virtv1.RunStrategyHalted {
```

Comment:
The logic in this function is really making my head spin.

It looks like if somehow runStrategy=halt but restartRequest != nil can result in an active VMI never being terminated. I'm really not sure.

I think this function needs refactoring.

Reply:
Fair. I added quite a bit of comments. Hopefully that clarifies what's going on.

As far as the "runStrategy=halt", I think you might have a point. I'll take a look.